File: abeln-hb-lattice-2e567b22ecd1/source/AADistr.cpp
#include "AADistr.hpp"
#include "AA.hpp"
#include "Lattice.hpp"
#include "Residue.hpp"
#include "Chain.hpp"

#include <string.h>
#include <iostream>
#include <fstream>
#include <stdio.h>
#include <iostream>
#include <iomanip>
#include <sstream>
#include <cstdlib>
#include <cmath>


AADistr::AADistr(){

}

void AADistr::init(string filename, AA * aai){
  aaInt =aai;
  numaa = AA::NUMAA;
  sumFreq=0;
  //// read file
  ifstream aaFile (filename.c_str());
  if (!aaFile.is_open()){
    cout << "could not open file: "<<filename.c_str()<<  endl;
    exit(1);
  }
  int i=0;


  char line[500];
  while(aaFile.getline(line,200)){
    
    
    
    //// split line into fields
    char aa_c[5];
    int freq;
    sscanf(line,"%3s%d",aa_c,&freq);
    string aa = string(aa_c);
    
    cout<< aa<<endl;
    //// check if aa order is same as in neteraction matrix
    if(aa != "HOH"){
      if( aaInt->stringtoAA(aa)!=i){
	cout<< "amino acid order in file "<<filename<<" does not match interaction matrix";
	exit(1);
      }
      freqPDB[i]=freq;
      sumFreq+=freq;
    }
    i++;
  }  
}

void AADistr::setLength(int length){
  seqLength=length;
  for(int i=0;i<numaa;i++){
    if(i != AA::WATER){ 
      freqExp[i]=freqPDB[i]*seqLength/sumFreq;
    }else{
      freqExp[i]=0;
    }
  }

}


void AADistr::setSequence(Lattice *l){
  int length=0;
  //// reset frequency table
  for(int i=0;i<AA::NUMAA;i++){
    freqSeq[i]=0;
  }
  //// 
  for(int nc=0;nc<l->nChains;nc++){
    Chain * ch = l->chains[nc];
    for (int n=0;n<ch->N;n++){
      Residue * res =ch->residues[n];
      freqSeq[res->aa]++;
      length++;
    }
  }
  setLength(length);
  systemDistance = getDistance();
}




double AADistr::getDistance(){
  double dist=0;
  for(int i=0;i<numaa;i++){
    if(i != AA::WATER){ 
      dist += (freqSeq[i] - freqExp[i])*(freqSeq[i] - freqExp[i]);
    }
  }
  return dist;
}


bool AADistr::check(){
  double calcD=getDistance();
  if(calcD != systemDistance ){
    cout <<"ERROR in design process accummulative distance:"<<systemDistance;
    cout <<" does not match calc distance:" <<calcD<<endl;
    return false;
  }
  return true;
}

double myround(double x,int decimal_places){
  double tmp = pow((double)10.0,(double)decimal_places)*x;
  return floor(tmp + 0.5)/  pow((double)10,(double)decimal_places);
}

string AADistr::toString(){
  std::stringstream outs;
  outs<<"system distance " <<systemDistance<<endl;
  outs<<"SEQUENCE"<<endl;
  double l=0;
  for(int aa=0;aa<AA::NUMAA;aa++){
    outs <<AA::int2aa[aa]<<" "<<freqSeq[aa]<<", ";
    l+= freqSeq[aa];
  }
  outs<< " length " <<l<< endl;
  outs<<"PDB"<<endl;
  double checksum =0;
  for(int aa=0;aa<AA::NUMAA;aa++){
    outs <<AA::int2aa[aa]<<" "<<myround(freqExp[aa],1)<<", ";
    checksum += freqExp[aa];
  }
  outs<< "SUM: "<<checksum <<endl;
  return outs.str();
} 

File: nelman-kmcLATTICEMODEL/source/AADistr.py
from typing import List 
from AA import AA, MAXAA 

class AADistr:
	def__init__(self):
		self.freqPDB = [0.0] * MAXAA
		self.freqExp = [0.0] * MAXAA
		self.freqSeq = [0.0] * MAXAA 
		self.seqLength = 0 
		self.numaa = 0 
		self.aaInt = None
		self.systemDistance = 0.0
		self.sumFreq = 0 

	def init(self, filename: str, aai: AA): 
		self.aaInt = aai
		self.numaa = AA.NUMAA 
		self.sumFreq = 0

		try:
			with open(filename, 'r') as aaFile: 
				i = 0
				for line in aaFile: 
					line = line.strip()
					if not line: 
						continue

					parts = line.split()
					if len(parts) < 2: 
						continue

					aa = parts[0].upper()
					freq = int(parts[1])
					print(aa)

					if aa != "HOH":
						if self.aaInt.stringtoAA(aa) != i:
							print(f"Amino acid order in file {filename} does not match interaction matrix")
							exit(1)
						self.freqPDB[i] = freq
						self.sumFreq += freq
					i += 1
		except IOError: 
			print(f"Could not open file: {filename}")
			exit(1)

	def setLength(self, length: int):
		self.seqLength = length
		for i in range(self.numaa):
			if i != AA.WATER:
				self.freqExp[i] = self.freqPDB[i] * self.seqLength / self.sumFreq
			else: self.freqExp[i] = 0 

	def setSequence(self, 1): 
		length = 0 
		for i in range(AA.NUMAA):
			self.freqSeq[i] = 0

		for nc in range(l.nChains): 
			ch = l.chains[nc]
			for n in range(ch.N): 
				res = ch.residues[n]
				self.freqSeq[res.aa] += 1
				length += 1

		self.setLength(length)
		self.systemDistance = self.getDistance()

	def getDistance(self) -> float:
		dist = 0.0 
		for i in range(self.numaa):
			if i != AA.WATER:
				dist += self.freqSeq[i] - self.freqExp[i]) ** 2 

		return dist

	def check(self) -> bool:
		calcD = self.getDistance()
		if calcD = self.systemDistance:
			print(f"ERROR in design process accumulative distance: {self.systemDistance} does not match calc distance: {calcD}")
			return False
		return True

	def substitute(self, aa1: int, aa2: int, dist: float):
		self.freqSeq[aa1] -= 1	
		self.freqSeq[aa2] += 1 
		self.systemDistance += dist

	def getDistanceSubst(self, aa1: int, aa2: int) -> float: 
		if aa1 = aa2:
			return 0
		return 2 * ((self.freqExp[aa1] - self.freqSeq[aa1]) - (self.freqExp[aa2] - self.freqSeq[aa2] - 1.0))

	def toString(self -> str: 
		import io 
		output = io.StringIO()

		output.write(f"System distance {self.systemDistance}\n") 
		output.write("SEQUENCE\n")
		1 = 0
		for aa in range(AA.NUMAA): 
			output.write(f"{AA.int2aa[aa]} {self.freqSeq[aa]}, ")
			1 += self.freqSeq[aa]
		output.write(f" Length {l}\n")

		output.write("PDB\n")
		checksum = 0 
		for aa in range(AA.NUMAA): 
			rounded = round(self.freqExp[aa], 1)
			output.write(f"{AA.int2aa[aa]} {rounded}, ")
			checksum += self.freqExp[aa]
		output.write(f"SUM: {checksum}\n"

		return output.getValue()
			     
